halotools/empirical_models/abunmatch/engines/bin_free_cam_kernel.pyx

```diff
@@ -110,11 +110,35 @@ def setup_initial_indices(int iy, int nwin, int npts):
     return iy, init_iy_low, init_iy_high
 
 
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+@cython.wraparound(False)
+cdef int _find_index(int[:] arr, int val):
```

Comment:
The `_find_index` function implements a blind lookup for a value in the correspondence indices array. The cython is clean (though watch out for the undeclared `index` variable on line 265), but I think the algorithm can be improved by taking advantage of the binary search that is already implemented, no? While the `correspondence_indx2` array is not sorted, the `sorted_cdf_values2` maintains a sorted order. If an original array of indices were defined at the very beginning of the engine, simply with `np.arange(npts2)`, couldn't that array be manipulated alongside `sorted_cdf_values2` to avoid the blind lookup? Naively, it seems this would be faster;  it could be that the performance hit from the additional manipulations wash out the improvement, I don't know, but it seems worth testing since the cythonized binary search machinery is already in place. Let me know whether this makes sense.

Reply:
Just to make sure I understand what you are suggesting:
In the same way that we have the `correspondence_indx2` array to keep track of how to unsort the `sorted_cdf_values2`, we can have a another array that keeps track of where things are in the `correspondence_indx2`? That would make this `_find_index` just a lookup in that map.

I **think** that this has roughly the same time complexity - when we update the correspondence indicies (to shift them along) we will need to update this array too and that would be roughly `O(len(nwin))` which is I think what we have now.

But I could well have missed something... I definitely don't fully understand all the internals.

(index `cdef`ed in f05be1c)

Reply:
You may be right, this might not be worth the trouble. After cleaning the cython of the remaining python interactions as discussed elsewhere, could you just do a simple timing test comparing your implementation against what is currently in master? If runtimes are comparable, then it means that further optimization is not worth spending time on.

Reply:
In the timing test, just make sure to use at least ~1e6 points